### PR TITLE
Apply fix for health API response to distinguish no master

### PR DESCRIPTION
**Pull request by mgodwan**
Signed-off-by: Mohit Godwani <mgodwan@amazon.com>

### Description
Add discovered_master attribute to cluster health API response
 
### Issues Resolved
https://github.com/opensearch-project/OpenSearch/issues/485
 
### Check List
- [x] New functionality includes testing.
  - [ ] All tests pass
  - [x] New tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/cluster/health/ClusterStateHealth.java
Comment (getsaurabh02) [inline server/src/main/java/org/opensearch/cluster/health/ClusterStateHealth.java:152]: I believe we should have Version Check here on input stream to allow backward compatibility and prevent serialisation issues.

File: server/src/main/java/org/opensearch/cluster/health/ClusterStateHealth.java
Comment (getsaurabh02) [inline server/src/main/java/org/opensearch/cluster/health/ClusterStateHealth.java:240]: Similar version check needed here on the output stream as well.

**Review by getsaurabh02 — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/opensearch/cluster/health/ClusterStateHealth.java
Comment (mgodwan) [reply server/src/main/java/org/opensearch/cluster/health/ClusterStateHealth.java]: Done.

File: server/src/main/java/org/opensearch/cluster/health/ClusterStateHealth.java
Comment (mgodwan) [reply server/src/main/java/org/opensearch/cluster/health/ClusterStateHealth.java]: Done.

**Review by getsaurabh02 — APPROVED** [review]

**Review by itiyamas — APPROVED** [review]

**Comment by mgodwan** [comment]
Can maintainers please help me merge this PR?

**Comment by itiyamas** [comment]
I have approved this PR. Once the checks pass, you will get a chance to merge this.

**Comment by itiyamas** [comment]
start gradle check

**Comment by mgodwan** [comment]
Need help with merging this PR. @saratvemulapalli Could you help with it?

**Comment by saratvemulapalli** [comment]
start gradle check

File: server/src/main/java/org/opensearch/cluster/health/ClusterStateHealth.java
Comment (saratvemulapalli) [inline server/src/main/java/org/opensearch/cluster/health/ClusterStateHealth.java:156]: Not sure I understand this correctly.
I know we have to initialize `hasDiscoveredMaster` but we are saying any version older than OpenSearch 1.0.0 has always discovered a master, even though that might not be true?

**Review by saratvemulapalli — APPROVED** [review]
The changes overall look good to me. 

Dropped a minor comment.

**Comment by itiyamas** [comment]
@mgodwan You have the required approvals. You can merge this too.

**Comment by mgodwan** [comment]
> @mgodwan You have the required approvals. You can merge this too.

@itiyamas Cannot merge by myself since only those with write access to this repository can merge pull requests.

File: server/src/main/java/org/opensearch/cluster/health/ClusterStateHealth.java
Comment (mgodwan) [reply server/src/main/java/org/opensearch/cluster/health/ClusterStateHealth.java:156]: The check is in place for backward compatibility since the older versions don't have this attribute in their Transport layer payload stream. 
Also, in the older versions, this value is not being populated in the API response, hence customers would not see this value either and they return 503 in case of master not discovered meaning this would not be invoked if master was not discovered.

Comment (saratvemulapalli) [reply server/src/main/java/org/opensearch/cluster/health/ClusterStateHealth.java:156]: Thats a good point, thanks for helping me understand.

**Pull request merged**
